### ISSUE TITLE
The model converted to coreml format always shows confidence 1.0

### ISSUE THREAD

**Issue by igor-tkc** [ultralytics/ultralytics#14192]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar bug report.

### YOLOv8 Component

Export

### Bug

I converted the custom trained yolov8 based model to coreml format and found that it always shows confident 1.0 in my iOS application.

See my CLI command:
`yolo export model=~/Downloads/model.pt format=coreml nms=True`

I also tried to convert without Non-Maximum Suppression (nms=False), but the model lost class labels.

Please, advise. Thanks

### Environment

ultralytics 8.2.48, Python 3.11, macOS 13.6.7(M1), iOS 17.5.1

### Are you willing to submit a PR?

- [ ] Yes I'd like to help by submitting a PR!

**Comment by glenn-jocher** [comment]
@igor-tkc hi there,

Thank you for bringing this issue to our attention. It seems like you're encountering an unusual behavior with the confidence scores after converting your YOLOv8 model to CoreML format.

To better assist you, could you please provide a minimal reproducible example? This will help us diagnose the issue more effectively. You can find guidelines on how to create one here: [Minimum Reproducible Example](https://docs.ultralytics.com/help/minimum_reproducible_example).

In the meantime, please ensure that you are using the latest version of the Ultralytics package. Sometimes, issues are resolved in newer releases, and updating might fix the problem.

Here's a quick checklist to verify:
1. **Update Ultralytics**: Ensure you have the latest version installed.
   ```bash
   pip install --upgrade ultralytics
   ```
2. **Export Command**: Your export command looks correct, but let's double-check it with the latest version:
   ```bash
   yolo export model=~/Downloads/model.pt format=coreml nms=True
   ```

If the issue persists, please share any additional details or code snippets that might help us reproduce the problem on our end.

Thank you for your cooperation, and we look forward to resolving this issue for you! 😊

**Comment by igor-tkc** [comment]
I tried to predict test image on both yolov8m.pt and converted yolov8m.mlpackage. See differences of confidence values.

![pt_dog_bike_car](https://github.com/ultralytics/ultralytics/assets/20066670/d5c383a7-b137-4ea0-9593-03237f439e49)
![coreml_dog_bike_car](https://github.com/ultralytics/ultralytics/assets/20066670/43bbb79b-220c-48eb-bc96-ad30553fa47d)

See snippet of CLI command and output of exporting to coreml. 
```
(.venv) bash-3.2$ yolo export model=yolov8m.pt format=coreml nms=True
Ultralytics YOLOv8.2.48 🚀 Python-3.11.1 torch-2.3.1 CPU (Apple M1)
YOLOv8m summary (fused): 218 layers, 25886080 parameters, 0 gradients, 78.9 GFLOPs

PyTorch: starting from 'yolov8m.pt' with input shape (1, 3, 640, 640) BCHW and output shape(s) (1, 84, 8400) (49.7 MB)
Torch version 2.3.1 has not been tested with coremltools. You may run into unexpected errors. Torch 2.2.0 is the most recent version that has been tested.

CoreML: starting export with coremltools 7.2...
Tuple detected at graph output. This will be flattened in the converted model.
...
CoreML Pipeline: starting pipeline with coremltools 7.2...
CoreML Pipeline: pipeline success
CoreML: export success ✅ 19.0s, saved as 'yolov8m.mlpackage' (49.6 MB)
```

The same confidence values I got in my Swift code for iOS:
```
for observation in results where observation is VNRecognizedObjectObservation {
            guard let objectObservation = observation as? VNRecognizedObjectObservation else {
                continue
            }
            let topLabelObservation = objectObservation.labels[0]
            print(topLabelObservation.confidence)
        }
```

**Comment by igor-tkc** [comment]
I also fixed inconsistent version of torch, but the behaviour is the same.
`Torch version 2.3.1 has not been tested with coremltools. You may run into unexpected errors. Torch 2.2.0 is the most recent version that has been tested.`

**Comment by glenn-jocher** [comment]
Thank you for providing the additional details and images. It looks like you've already addressed the version inconsistency with PyTorch, which is great.

To further investigate this issue, let's ensure a few more things:

1. **CoreMLTools Compatibility**: Since `coremltools` has been tested up to PyTorch 2.2.0, please make sure you are using this version to avoid any unexpected behavior. You can downgrade PyTorch using:
   ```bash
   pip install torch==2.2.0
   ```

2. **Model Export with Correct Settings**: Ensure that the export command includes the necessary arguments for proper conversion:
   ```bash
   yolo export model=~/Downloads/model.pt format=coreml nms=True
   ```

3. **Verify Predictions in Python**: Before deploying to iOS, verify the CoreML model predictions in Python to ensure the issue isn't specific to the iOS environment. You can use the `coremltools` library to load and test the model:
   ```python
   import coremltools as ct
   from PIL import Image
   import numpy as np

   # Load the CoreML model
   model = ct.models.MLModel('yolov8m.mlpackage')

   # Load and preprocess the image
   image = Image.open('path/to/your/image.jpg')
   input_data = np.array(image.resize((640, 640))).astype(np.float32)

   # Make predictions
   results = model.predict({'image': input_data})
   print(results)
   ```

If the confidence values are correct in Python but not in iOS, the issue might be related to how the model is being used or interpreted in the iOS environment.

4. **Check iOS Implementation**: Ensure that the iOS code correctly interprets the model's output. Sometimes, post-processing steps might be necessary to correctly extract and interpret confidence scores.

If the issue persists, please provide a minimal reproducible example as it will help us diagnose the problem more effectively. You can find guidelines on creating one here: [Minimum Reproducible Example](https://docs.ultralytics.com/help/minimum_reproducible_example).

Thank you for your patience, and we look forward to resolving this issue for you! 😊
